### ISSUE TITLE
Commit for B2B-2404 causes issues with GraphQL Product Data Provider

### ISSUE THREAD

**Issue by pmzandbergen** [magento/magento2#37405]
### Summary

Starting from version 2.4.6 the Product Data Provider filters on the Visibility, this is not always desired (depending on the resolver). See commit:
https://github.com/magento/magento2/commit/8681d30d2517a47baba01be70dc08031a0ab2671#diff-f94c66fe63952d9bf9efbd6d9cb69ddcb479df56dde42b9d9327a7bf4a9446abR137

File:
https://github.com/magento/magento2/blob/2.4-develop/app/code/Magento/CatalogGraphQl/Model/Resolver/Products/DataProvider/Deferred/Product.php#L138

Filtering (in this case on Visibility) is (in my opinion) something which should be handled by the Resolver, not the Data Provider.

### Examples

There are multiple resolvers in 3rd party modules using the Product Data Provider. These modules expect the products to be returned, no matter their visibility. Please note that disabled or unassigned products is another story, those are expected to be filtered.

### Proposed solution

Review if the filtering is desired and if so, move it to the Resolver.

### Additional Information
For further commentary on this issue, this also impacts bundled products.

If you add a simple product that is `visiblity - IN_CATALOG_ONLY` that is not in any category on the site, then this will cause the `product` node of the `BundleItemOption` to return `null` since the [underlying join](https://github.com/magento/magento2/blob/2.4-develop/app/code/Magento/Catalog/Model/ResourceModel/Product/Collection.php#L2070) is done against `catalog_category_product_index_storeX` which will not contain products if they are not assigned to any category.

This is not expected behavior because there may be a product that you want to sell as part of a bundle that isn't a part of any specific category since there's no need to list it anywhere.

### Triage and priority

- [ ] Severity: **S0** _- Affects critical data or functionality and leaves users without workaround._
- [ ] Severity: **S1** _- Affects critical data or functionality and forces users to employ a workaround._
- [X] Severity: **S2** _- Affects non-critical data or functionality and forces users to employ a workaround._
- [ ] Severity: **S3** _- Affects non-critical data or functionality and does not force users to employ a workaround._
- [ ] Severity: **S4** _- Affects aesthetics, professional look and feel, “quality” or “usability”._

**Comment by joanhe** [comment]
@pmzandbergen This change is done for fixing bugs causing by over fetching ignoring context. Do you see any use case that has conflict with this change? If so, please add to the issue report, so we can evaluate it.

**Comment by ananth-iyer** [comment]
Hi @joanhe on line 139, it is causing a major issue. I am getting grouped product and its child products by GraphQL. It returns grouped product details but child products are empty.

Tagging @pmzandbergen

Issue on Magento 2.4.6 and 2.4.6-p1
PHP 8.1

https://github.com/magento/magento2/blob/ca30c47787db932b9431e42226bb918e95c932d9/app/code/Magento/CatalogGraphQl/Model/Resolver/Products/DataProvider/Deferred/Product.php#L134-L141

On line 139, if you change the parameter from `true` to `false` then it returns all child product properly.

![query-25 07 2023_18 42 23_REC](https://github.com/magento/magento2/assets/24892273/8c2bafcc-f3a4-44a5-b3bb-6d1252b9d658)

![25 07 2023_18 40 57_REC](https://github.com/magento/magento2/assets/24892273/3eb861ac-2959-4791-b739-2f322d4e216b)

Please check this again.

**Comment by ananth-iyer** [comment]
@engcom-Hotel tagging for visibility

**Comment by ananth-iyer** [comment]
@engcom-Dash

**Comment by engcom-November** [comment]
Hello @pmzandbergen,
Thank you for report and collaboration!
Thank you @ananth-iyer for your input!

We have tried to reproduce this issue on 2.4-develop CE and B2B.
We are getting expected results.
Please take a look at the screenshot below:
![image](https://github.com/magento/magento2/assets/60198231/48916344-b46c-4ffa-8011-0865897dfc0f)
We are able to get the child product properly from the below query.
```
{
  products(filter:{
    url_key:{
      eq:"a-bag1"
    }
  }) {
    items {
      url_key
      sku
         price_range {
             minimum_price {
                    final_price {
                      value
                      currency
                    }
              
              discount {
                percent_off
                amount_off
              }
            
              regular_price {
                value
                currency
              }
              
                  }
          
                  maximum_price {
                    final_price {
                      value
                      currency
                    }
                    regular_price {
                      value
                      currency
                    }
                    discount {
                      percent_off
                      amount_off
                    }
                  }
        }
   ... on GroupedProduct {
       items {
           position
           qty
           product {
               uid
               sku
               name
               only_x_left_in_stock
               price_range {
                   maximum_price {
                       final_price {
                           currency
                           value
                       }
                       regular_price {
                           currency
                           value
                       }
                   }
                   minimum_price {
                       final_price {
                           currency
                           value
                       }
                       regular_price {
                           currency
                           value
                       }
                   }
               }
           }
       }
   }
      
    }
  }
}
``` 
Please let us know if we are missing something.

Thank you.

**Comment by pmzandbergen** [comment]
> We have tried to reproduce this issue on 2.4-develop CE and B2B. We are getting expected results. 

That is correct, on a native Magento installation you will get the expected results. Please see the original description:
_"Filtering (in this case on Visibility) is (in my opinion) something which should be handled by the Resolver, not the Data Provider."_

Its more a discussion about resolvers VS providers. Changing the behavior of the provider affected some 3rd party modules, since they didn't expect the provider to apply additional filters on its own. Isn't that something which should be handled by the resolver?

**Comment by engcom-November** [comment]
Hello @pmzandbergen,

Thank you for your quick response!

As the issue is not present in the core-magento library, but some 3rd party modules, please let us know in which modules you are getting the issue, this will help us to dig deep into the issue.

Thank you.

**Comment by pmzandbergen** [comment]
@engcom-November please have a look at the comment from @ananth-iyer: https://github.com/magento/magento2/issues/37405#issuecomment-1649827128

It seems this change does cause issues in the Magento core. Even if it didn't the original issue shouldn't be fixed by changing the Data Provider in my opinion.

- Resolver: Its task is to resolve data, for example: _"The first 100 products with the price between X and Y ordered by price"_ or _"All child products for grouped product XYZ"_
- Data Provider: Convert the (to be) resolved products based on the requested attributes (depending on the GraphQL query)

With this change its the Data Provider applying filters, thats not its role (_except for, maybe, filtering disabled products_).

**Comment by engcom-November** [comment]
Hello @pmzandbergen,

Thank you for the response, and we agree with your point https://github.com/magento/magento2/issues/37405#issuecomment-1698600535.

Resolver should add filters and not data provider, but as we can see [here](https://github.com/magento/magento2/blob/ca30c47787db932b9431e42226bb918e95c932d9/app/code/Magento/CatalogGraphQl/Model/Resolver/Products/DataProvider/Deferred/Product.php#L134) data provider is applying the filter. 
Hence this issue can be confirmed.

Thank you.

**Comment by damienwebdev** [comment]
For further commentary on this issue, this also impacts bundled products.

If you add a simple product that is `visiblity - IN_CATALOG` only that is not in any category on the site, then this will cause the `product` node of the `BundleItemOption` to return `null` since the [underlying join](https://github.com/magento/magento2/blob/2.4-develop/app/code/Magento/Catalog/Model/ResourceModel/Product/Collection.php#L2070) is done against `catalog_category_product_index_storeX` which will not contain products if they are not assigned to any category.

**Comment by engcom-Hotel** [comment]
Thanks @pmzandbergen for more information on this. We have added the same in `Additional Information` of the main description.
